src/mlir/ir/attribute/dense_string.rs

```diff
@@ -0,0 +1,178 @@
+/// - `mlirElementsAttrGetNumElements`
+/// - `mlirDenseElementsAttrGetStringValue`
+/// - `mlirDenseElementsAttrStringGet`
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct DenseStringAttributeRef {
+    _prevent_external_instantiation: PhantomData<()>,
+    raw: MlirAttribute,
+}
+
+impl_unowned_mlir_value!(no_refs, DenseStringAttributeRef, MlirAttribute);
+impl_attribute_variant!(DenseStringAttributeRef, mlirAttributeIsADenseElements);
+
+impl DenseStringAttributeRef {
+    /// Constructs a new dense array of strings attribute with the provided values.
+    ///
+    /// # Arguments
+    /// * `context` - The context that should own the attribute.
+    /// * `values` - The strings to hold in the attribute.
+    /// * `string_type` - The string type to be parsed to an MlirType.
+    ///
+    /// # Returns
+    /// Returns a reference to a new [`DenseStringAttributeRef`] instance.
+    pub fn new<'a>(
+        context: &ContextRef,
+        values: &[StringRef],
+        string_type: &str,
+    ) -> &'a DenseStringAttributeRef {
+        let shape: [i64; 1] = [values.len() as i64];
+
+        let shaped_type = unsafe {
+            mlirRankedTensorTypeGet(
+                1,
+                shape.as_ptr(),
+                TypeRef::parse(&context, string_type).unwrap().to_raw(),
+                mlirAttributeGetNull(),
+            )
+        };
+
+        unsafe {
+            Self::from_raw(mlirDenseElementsAttrStringGet(
+                shaped_type,
+                values.len() as isize,
+                values.as_ptr() as *mut MlirStringRef,
+            ))
+        }
+    }
+
+    /// # Returns
+    /// Returns the length of the array.
+    pub fn len(&self) -> isize {
+        unsafe { mlirElementsAttrGetNumElements(self.to_raw()) as isize }
+    }
+
+    /// # Returns
+    /// Returns whether the array is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Gets the element at the provided index, verifying that the index is within bounds.
+    ///
+    /// # Arguments
+    /// * `index` - The index of the element to get.
+    ///
+    /// # Returns
+    /// Returns the StringRef at the provided index.
+    pub fn get(&self, index: isize) -> StringRef {
+        assert!(index < self.len().try_into().unwrap());
+        unsafe {
+            let string_ref = mlirDenseElementsAttrGetStringValue(self.to_raw(), index);
+            StringRef::from_raw(string_ref)
+        }
+    }
+
+    /// # Returns
+    /// Returns a named attribute with the provided name.
+    pub fn with_name(&self, name: &str) -> NamedAttribute {
+        let identifier = IdentifierRef::new(&self.context(), name);
+        unsafe { NamedAttribute::from_raw(identifier.to_raw(), self.to_raw()) }
+    }
+
+    /// # Returns
+    /// Returns a reference to the context that owns the attribute.
+    pub fn context(&self) -> &ContextRef {
+        unsafe { ContextRef::from_raw(mlirAttributeGetContext(self.to_raw())) }
+    }
```

Comment:
Since you treat this as an attribute variant, I don't think you need these functions defined here manually? They should be accessible automatically via the `Deref` impl provided by the `impl_attribute_variant` macro.